# **A. Persiapan & Import Library**

## **1. Connect Google Drive**

In [295]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


## **2. Import Library**

In [296]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import re
import seaborn as sns

## **3. Memindai Lokasi Dataset di Google Drive**

In [297]:
file_path = '/content/drive/MyDrive/2026GoogleColab_DataMining/Tugas1/Dataset/df_arabica.csv'

## **4. Membaca Dataset**

In [298]:
df = pd.read_csv(file_path)

print("Dataset berhasil dibaca!")
print(f"Jumlah baris: {df.shape[0]} | Jumlah kolom: {df.shape[1]}")

Dataset berhasil dibaca!
Jumlah baris: 207 | Jumlah kolom: 41


Dataset yang digunakan memiliki 207 baris dan 41 kolom

# **B. Eksplorasi Data Awal**

## **1. Menampilkan 10 Dataset Teratas**

In [299]:
print("---  10 Baris Pertama Dataset ---")
display(df.head(10))

---  10 Baris Pertama Dataset ---


,Unnamed: 0,ID,Country of Origin,Farm Name,Lot Number,Mill,ICO Number,Company,Altitude,Region,...,Total Cup Points,Moisture Percentage,Category One Defects,Quakers,Color,Category Two Defects,Expiration,Certification Body,Certification Address,Certification Contact
0,0,0,Colombia,Finca El Paraiso,CQU2022015,Finca El Paraiso,NaN,Coffee Quality Union,1700-1930,"Piendamo,Cauca",...,89.33,11.8,0,0,green,3,"September 21st, 2023",Japan Coffee Exchange,"〒413-0002 静岡県熱海市伊豆山１１７３−５８ 1173-58 Izusan, Ata...",松澤 宏樹 Koju Matsuzawa - +81(0)9085642901
1,1,1,Taiwan,Royal Bean Geisha Estate,"The 2022 Pacific Rim Coffee Summit,T037",Royal Bean Geisha Estate,NaN,Taiwan Coffee Laboratory,1200,Chiayi,...,87.58,10.5,0,0,blue-green,0,"November 15th, 2023",Taiwan Coffee Laboratory 台灣咖啡研究室,"QAHWAH CO., LTD 4F, No. 225, Sec. 3, Beixin Rd...","Lin, Jen-An Neil 林仁安 - 886-289116612"
2,2,2,Laos,OKLAO coffee farms,"The 2022 Pacific Rim Coffee Summit,LA01",oklao coffee processing plant,NaN,Taiwan Coffee Laboratory,1300,Laos Borofen Plateau,...,87.42,10.4,0,0,yellowish,2,"November 15th, 2023",Taiwan Coffee Laboratory 台灣咖啡研究室,"QAHWAH CO., LTD 4F, No. 225, Sec. 3, Beixin Rd...","Lin, Jen-An Neil 林仁安 - 886-289116612"
3,3,3,Costa Rica,La Cumbre,CQU2022017,La Montana Tarrazu MIll,NaN,Coffee Quality Union,1900,"Los Santos,Tarrazu",...,87.17,11.8,0,0,green,0,"September 21st, 2023",Japan Coffee Exchange,"〒413-0002 静岡県熱海市伊豆山１１７３−５８ 1173-58 Izusan, Ata...",松澤 宏樹 Koju Matsuzawa - +81(0)9085642901
4,4,4,Colombia,Finca Santuario,CQU2023002,Finca Santuario,NaN,Coffee Quality Union,1850-2100,"Popayan,Cauca",...,87.08,11.6,0,2,yellow-green,2,"March 5th, 2024",Japan Coffee Exchange,"〒413-0002 静岡県熱海市伊豆山１１７３−５８ 1173-58 Izusan, Ata...",松澤 宏樹 Koju Matsuzawa - +81(0)9085642901
5,5,5,Guatemala,La Colina,"The 2022 Pacific Rim Coffee Summit,GT02",Dinámica Café,NaN,Taiwan Coffee Laboratory,1668,Chimaltenango,...,87.00,10.7,0,0,green,2,"November 15th, 2023",Taiwan Coffee Laboratory 台灣咖啡研究室,"QAHWAH CO., LTD 4F, No. 225, Sec. 3, Beixin Rd...","Lin, Jen-An Neil 林仁安 - 886-289116612"
6,6,6,Taiwan,野牡丹咖啡莊園 Melastoma Coffee Estate,"The 2022 Pacific Rim Coffee Summit,T034",野牡丹咖啡,NaN,Taiwan Coffee Laboratory,1250,Chiayi,...,86.92,9.1,0,0,green,0,"November 15th, 2023",Taiwan Coffee Laboratory 台灣咖啡研究室,"QAHWAH CO., LTD 4F, No. 225, Sec. 3, Beixin Rd...","Lin, Jen-An Neil 林仁安 - 886-289116612"
7,7,7,Taiwan,七彩琉璃咖啡莊園 Chi Tsai Liu Li Ecological Farm,"The 2022 Pacific Rim Coffee Summit,T050",七彩琉璃咖啡莊園,NaN,Taiwan Coffee Laboratory,1200,Chiayi,...,86.75,10.0,0,0,yellow green,1,"November 15th, 2023",Taiwan Coffee Laboratory 台灣咖啡研究室,"QAHWAH CO., LTD 4F, No. 225, Sec. 3, Beixin Rd...","Lin, Jen-An Neil 林仁安 - 886-289116612"
8,8,8,Taiwan,亮軒咖啡莊園 Liang Xuan Coffee Farm,"The 2022 Pacific Rim Coffee Summit,T018",亮軒咖啡莊園,NaN,Taiwan Coffee Laboratory,1250,Chiayi,...,86.67,10.8,0,0,greenish,0,"November 15th, 2023",Taiwan Coffee Laboratory 台灣咖啡研究室,"QAHWAH CO., LTD 4F, No. 225, Sec. 3, Beixin Rd...","Lin, Jen-An Neil 林仁安 - 886-289116612"
9,9,9,"Tanzania, United Republic Of",MASHIMA AMCOS,CN 4127230034/4189230113,GOURMET COFFEE MILL,033/DE/503/002 and 033/DE/268/002,DORMAN (T) LIMITED,1400-1700,KILIMANJARO,...,86.50,11.0,0,0,greenish,0,"February 1st, 2024",Kenya Coffee Traders Association,"P.O. Box 646 00100Nairobi, Kenya, Denniss Prit...",Benson Kibicho Kamau - +254 786 821 621


## **2. Informasi tipe data dan non-null count**

In [300]:
print("\n--- Ringkasan Tipe Data & Struktur ---")
df.info()


--- Ringkasan Tipe Data & Struktur ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 207 entries, 0 to 206
Data columns (total 41 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             207 non-null    int64  
 1   ID                     207 non-null    int64  
 2   Country of Origin      207 non-null    object 
 3   Farm Name              205 non-null    object 
 4   Lot Number             206 non-null    object 
 5   Mill                   204 non-null    object 
 6   ICO Number             75 non-null     object 
 7   Company                207 non-null    object 
 8   Altitude               206 non-null    object 
 9   Region                 205 non-null    object 
 10  Producer               206 non-null    object 
 11  Number of Bags         207 non-null    int64  
 12  Bag Weight             207 non-null    object 
 13  In-Country Partner     207 non-null    object 
 14  Harvest Year      

Dataset yang digunakan memiliki 13 kolom dengan tipe data float64 & 6 kolom int64, serta 22 kolom object.

## **3. Memeriksa jumlah nilai kosong (missing values) per kolom**

In [301]:
print("\n--- Jumlah Missing Value per Kolom ---")
missing_values = df.isnull().sum()
print(missing_values[missing_values > 0])


--- Jumlah Missing Value per Kolom ---
Farm Name              2
Lot Number             1
Mill                   3
ICO Number           132
Altitude               1
Region                 2
Producer               1
Variety                6
Processing Method      5
dtype: int64


Dataset yang digunakan memiliki beberapa missing value dibeberapa kolom yang ada.

## **4. Memeriksa baris terduplikat**

In [302]:
print("\n--- Cek Baris Duplikat ---")
print(f"Jumlah baris duplikat: {df.duplicated().sum()}")


--- Cek Baris Duplikat ---
Jumlah baris duplikat: 0


Dataset yang digunakan tidak memililki data yang duplikat.

# **C. Menghapus Kolom Duplikat (Redudansi)**

## **1. Menghapus kolom ID yang tidak digunakan**

In [303]:
cols_to_drop = [
    col for col in ['Unnamed: 0', 'ICO Number'] if col in df.columns
]
df.drop(columns=cols_to_drop, inplace=True)

Menghapus kolom Unnamed karena fungsinya sama dengan kolom ID dan ICO Number dihapus karena data yang hilang dari kolom tersebut lebih dari setengah jumlah data.

# **D. Standarisasi Tipe Data Numerik (Bag Weight & Altitude)**

## **1. Ekstraksi angka berat bag dari teks '35 kg' -> 35.0**

In [304]:
def convert_bag_weight(val):
  if pd.isna(val) or str(val).lower() == 'unknown':
    return np.nan
  val_str = str(val).lower().strip()

  numbers = re.findall(r'\d+(?:\.\d+)?', val_str)
  if not numbers:
    return np.nan

  weight = float(numbers[0])

  #Konversi ke KG
  if 'lbs' in val_str or 'pound' in val_str:
    weight = weight * 0.45359237
  elif 'g' in val_str and 'kg' not in val_str:
    weight = weight / 1000.0

  return weight


if 'Bag Weight' in df.columns:
  df['Bag Weight'] = df['Bag Weight'].apply(convert_bag_weight)

print('Standarisasi Bag Weight selesai!')
display(df[['Bag Weight']].head())

Standarisasi Bag Weight selesai!


,Bag Weight
0,35.0
1,80.0
2,25.0
3,22.0
4,24.0


Standarisasi data pada kolom Bag Weight yang memiliki teks seperti "35 kg" menjadi angka utamanya saja dengan menjadikannya nilai float 35.0.

## **2. Konversi Altitude yang berupa range ('1700-1930') menjadi rerata numerik**

In [305]:
def extract_altitude_mean(val):
    if pd.isna(val):
        return np.nan

    numbers = [float(s) for s in re.findall(r'\d+', str(val))]
    if len(numbers) >= 2:
        return sum(numbers[:2]) / 2
    elif len(numbers) == 1:
        return numbers[0]
    return np.nan

if 'Altitude' in df.columns:
    df['Altitude'] = df['Altitude'].apply(extract_altitude_mean)

print("Standarisasi Altitude selesai!")
display(df[['Altitude']].head())

Standarisasi Altitude selesai!


,Altitude
0,1815.0
1,1200.0
2,1300.0
3,1900.0
4,1975.0


Standarisasi data pada kolom Altitude yang memiliki data rentan seperti 1700-1930, kemudian mengambil data tengah-tengah atau mean, dari yang 1700-1930 nilai tengahnya 1815.

# **E. Standarisasi Tipe Data Tanggal (Grading Date & Expiration)**

## **1. Membuat Fungsi untuk Standarisasi, Pembersihan, dan Konversi ke pd.to_datetime**

In [306]:
date_cols = ['Grading Date', 'Expiration']

for col in date_cols:
  if col in df.columns:
    df[col] = pd.to_datetime(df[col], errors='coerce', format='mixed')

print('Format tanggal berhasil diubah ke datetime64!')
display(df[date_cols].dtypes)
display(df[date_cols].head())

Format tanggal berhasil diubah ke datetime64!


,0
Grading Date,datetime64[ns]
Expiration,datetime64[ns]


,Grading Date,Expiration
0,2022-09-21,2023-09-21
1,2022-11-15,2023-11-15
2,2022-11-15,2023-11-15
3,2022-09-21,2023-09-21
4,2023-03-06,2024-03-05


Standarisasi kolom Granding Date dan Expiration yang datanya merupakan tanggal dari st, nd, rd, th. Dijadikan tanggal dengan format YYYY-MM-DD

# **F. Pembersihan Aksara Non-Latin**

Standarisasi kolom yang datanya memiliki aksara non-Latin, dengan menyisakan datanya berupa alfabet Latin

## **1. Membuat Fungsi untuk Menghapus Aksara Non-Latin dan Mempertahankan Karakter Latin**

In [307]:
def remove_non_latin(text):
  if pd.isna(text) or not isinstance(text, str):
    return text

  # Menghapus aksara non-Latin
  cleaned_text = re.sub(r'[^\x00-\x7F]+', '', text)
  cleaned_text = ' '.join(cleaned_text.split())

  # Jika string jadi kosong setelah dihapus, ganti ke Unknown
  return cleaned_text if cleaned_text else 'Unknown'


# Menerapkan pada semua kolom teks (object)
text_cols = df.select_dtypes(include=['object']).columns
for col in text_cols:
  df[col] = df[col].apply(remove_non_latin)

print('Pembersihan karakter non-Latin selesai!')

display(df[text_cols].head())

Pembersihan karakter non-Latin selesai!


,Country of Origin,Farm Name,Lot Number,Mill,Company,Region,Producer,In-Country Partner,Harvest Year,Owner,Variety,Status,Processing Method,Color,Certification Body,Certification Address,Certification Contact
0,Colombia,Finca El Paraiso,CQU2022015,Finca El Paraiso,Coffee Quality Union,"Piendamo,Cauca",Diego Samuel Bermudez,Japan Coffee Exchange,2021 / 2022,Coffee Quality Union,Castillo,Completed,Double Anaerobic Washed,green,Japan Coffee Exchange,"413-0002 1173-58 Izusan, Atami, Shizuoka, 413-...",Koju Matsuzawa - +81(0)9085642901
1,Taiwan,Royal Bean Geisha Estate,"The 2022 Pacific Rim Coffee Summit,T037",Royal Bean Geisha Estate,Taiwan Coffee Laboratory,Chiayi,Unknown,Taiwan Coffee Laboratory,2021 / 2022,Taiwan Coffee Laboratory,Gesha,Completed,Washed / Wet,blue-green,Taiwan Coffee Laboratory,"QAHWAH CO., LTD 4F, No. 225, Sec. 3, Beixin Rd...","Lin, Jen-An Neil - 886-289116612"
2,Laos,OKLAO coffee farms,"The 2022 Pacific Rim Coffee Summit,LA01",oklao coffee processing plant,Taiwan Coffee Laboratory,Laos Borofen Plateau,WU TAO CHI,Taiwan Coffee Laboratory,2021 / 2022,Taiwan Coffee Laboratory,Java,Completed,Semi Washed,yellowish,Taiwan Coffee Laboratory,"QAHWAH CO., LTD 4F, No. 225, Sec. 3, Beixin Rd...","Lin, Jen-An Neil - 886-289116612"
3,Costa Rica,La Cumbre,CQU2022017,La Montana Tarrazu MIll,Coffee Quality Union,"Los Santos,Tarrazu",Santa Maria de Dota,Japan Coffee Exchange,2022,Coffee Quality Union,Gesha,Completed,Washed / Wet,green,Japan Coffee Exchange,"413-0002 1173-58 Izusan, Atami, Shizuoka, 413-...",Koju Matsuzawa - +81(0)9085642901
4,Colombia,Finca Santuario,CQU2023002,Finca Santuario,Coffee Quality Union,"Popayan,Cauca",Camilo Merizalde,Japan Coffee Exchange,2022,Coffee Quality Union,Red Bourbon,Completed,"Honey,Mossto",yellow-green,Japan Coffee Exchange,"413-0002 1173-58 Izusan, Atami, Shizuoka, 413-...",Koju Matsuzawa - +81(0)9085642901


# **G. Pembersihan Spasi & Deduplikasi Data**

## **1. Membersihkan spasi berlebih di awal dan akhir teks pada semua kolom string**

In [308]:
object_cols = df.select_dtypes(include=['object']).columns
for col in object_cols:
    df[col] = df[col].astype(str).str.strip()

## **2. Cek dan hapus data duplikat jika ada**

In [309]:
initial_count = len(df)
df.drop_duplicates(inplace=True)
final_count = len(df)

print(
    f" Bersihkan spasi selesai! Baris duplikat yang dihapus: {initial_count - final_count}"
)

 Bersihkan spasi selesai! Baris duplikat yang dihapus: 0


# **H. Ringkasan Akhir & Export Dataset Clean**

## **1. Ringkasan info dataset yang sudah bersih**

In [310]:
print("--- Info Dataset Hasil Cleansing ---")
df.info()
display(df.head())

--- Info Dataset Hasil Cleansing ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 207 entries, 0 to 206
Data columns (total 39 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   ID                     207 non-null    int64         
 1   Country of Origin      207 non-null    object        
 2   Farm Name              207 non-null    object        
 3   Lot Number             207 non-null    object        
 4   Mill                   207 non-null    object        
 5   Company                207 non-null    object        
 6   Altitude               206 non-null    float64       
 7   Region                 207 non-null    object        
 8   Producer               207 non-null    object        
 9   Number of Bags         207 non-null    int64         
 10  Bag Weight             207 non-null    float64       
 11  In-Country Partner     207 non-null    object        
 12  Harvest Year           207 

,ID,Country of Origin,Farm Name,Lot Number,Mill,Company,Altitude,Region,Producer,Number of Bags,...,Total Cup Points,Moisture Percentage,Category One Defects,Quakers,Color,Category Two Defects,Expiration,Certification Body,Certification Address,Certification Contact
0,0,Colombia,Finca El Paraiso,CQU2022015,Finca El Paraiso,Coffee Quality Union,1815.0,"Piendamo,Cauca",Diego Samuel Bermudez,1,...,89.33,11.8,0,0,green,3,2023-09-21,Japan Coffee Exchange,"413-0002 1173-58 Izusan, Atami, Shizuoka, 413-...",Koju Matsuzawa - +81(0)9085642901
1,1,Taiwan,Royal Bean Geisha Estate,"The 2022 Pacific Rim Coffee Summit,T037",Royal Bean Geisha Estate,Taiwan Coffee Laboratory,1200.0,Chiayi,Unknown,1,...,87.58,10.5,0,0,blue-green,0,2023-11-15,Taiwan Coffee Laboratory,"QAHWAH CO., LTD 4F, No. 225, Sec. 3, Beixin Rd...","Lin, Jen-An Neil - 886-289116612"
2,2,Laos,OKLAO coffee farms,"The 2022 Pacific Rim Coffee Summit,LA01",oklao coffee processing plant,Taiwan Coffee Laboratory,1300.0,Laos Borofen Plateau,WU TAO CHI,19,...,87.42,10.4,0,0,yellowish,2,2023-11-15,Taiwan Coffee Laboratory,"QAHWAH CO., LTD 4F, No. 225, Sec. 3, Beixin Rd...","Lin, Jen-An Neil - 886-289116612"
3,3,Costa Rica,La Cumbre,CQU2022017,La Montana Tarrazu MIll,Coffee Quality Union,1900.0,"Los Santos,Tarrazu",Santa Maria de Dota,1,...,87.17,11.8,0,0,green,0,2023-09-21,Japan Coffee Exchange,"413-0002 1173-58 Izusan, Atami, Shizuoka, 413-...",Koju Matsuzawa - +81(0)9085642901
4,4,Colombia,Finca Santuario,CQU2023002,Finca Santuario,Coffee Quality Union,1975.0,"Popayan,Cauca",Camilo Merizalde,2,...,87.08,11.6,0,2,yellow-green,2,2024-03-05,Japan Coffee Exchange,"413-0002 1173-58 Izusan, Atami, Shizuoka, 413-...",Koju Matsuzawa - +81(0)9085642901


Melakukan pengecekan ulang dataset yang sudah dibersihkan

## **2. Simpan dataset hasil pembersihan ke Google Drive**

In [311]:
output_file = (
    '/content/drive/MyDrive/2026GoogleColab_DataMining/Tugas1/2318039_df_arabica_cleaned.csv'
)
df.to_csv(output_file, index=False)

print(f"\n Dataset bersih berhasil disimpan ke: {output_file}")


 Dataset bersih berhasil disimpan ke: /content/drive/MyDrive/2026GoogleColab_DataMining/Tugas1/2318039_df_arabica_cleaned.csv
